### ISSUE TITLE
Support for (python) plotly figures

### ISSUE THREAD

**Issue by mwouts** [rstudio/reticulate#269]
Python plotly figures in Rmd documents appear as dictionaries. For instance,
```python
import plotly
import plotly.graph_objs as go
fig = {'data':[go.Scatter(x=[1,2,3,4],y=[1,4,9,16])],
       'layout':go.Layout(title='Plot generated in python')}
plotly.offline.iplot(fig)
```
yields
```
## {'application/vnd.plotly.v1+json': {'data': [{'type': 'scatter', 'x': [1, 2, 3, 4], 'y': [1, 4, 9, 16]}], 'layout': {'title': 'Plot generated in python'}}}
```

The dictionary seems to contain the data required to plot the figure in R. Could we display that kind of output as a plotly widget ?

An example for this is at https://github.com/mwouts/notebooks/blob/master/Python_plotly_in_R_notebook.Rmd

**Comment by jjallaire** [comment]
I think that to handle this correctly we'd need to support Jupyter Widgets: https://ipywidgets.readthedocs.io/en/latest/embedding.html

**Comment by mwouts** [comment]
Hello @jjallaire, I'm coming back to this subject, and also to that [one](https://github.com/rstudio/reticulate/issues/281), which is closely related. 

I think it should be feasible to
- capture ipython's rich outputs
- save plain images and PDFs to files, as currently done by `eng_python_matplotlib_show`
- capture python's json data for plotly plots, and emulate a R plotly plot, as in that [minimal working example](https://github.com/mwouts/notebooks/blob/master/Python_plotly_in_R_notebook.Rmd)

I could be working on a proof of concept, but... I'm not sure were to start ! Could you please give a few hints on how the plots are currently being captured ? I've seen and understood the call to `matplotlib.savefig` at https://github.com/rstudio/reticulate/blob/master/R/knitr-engine.R#L203, but I've not been able to find out how the library identifies that a plot should be captured.

**Comment by mwouts** [comment]
Thanks @kevinushey, that's clearer now! I'll try and see what I can do.

**Comment by dfrail24** [comment]
Has anyone found a good solution for this issue?

**Comment by mwouts** [comment]
Hello @dfrail24 , well I finally used the following approach:
- save the plotly graph made in Python as a JSON string
- read the JSON plot in R and plot it in R using R's plotly.

This way I was able to do the plot in Python, and use R Markdown to generate HTML slides. See the code at https://github.com/mwouts/jupytext_pyparis_2018/blob/master/slides/Jupytext.Rmd#L129-L180 and the resulting [HTML slides](https://mwouts.github.io/jupytext_pyparis_2018/slides/Jupytext.html#9).

**Comment by chitzinwin** [comment]
This discussion and [the other ](https://github.com/rstudio/rmarkdown/issues/1370) helped me a lot. Technically, I extended the method @dfrail24 suggested with tempfile(). @mwouts, I still agreed with your solution if we can extract dict from python object to feed it in r function. My case is dealing with html output right from map widget of [arcgis python api](https://developers.arcgis.com/python/guide/using-the-map-widget/)

```{python}
from arcgis.gis import GIS
gis = GIS()
`
```{r}
py$map1=gis$map('Grand Rapids, MI', 14)
py$map1$add_layer(gis$content$get("ea7bedd4d9a94e49bbf1b9207bc03d0c"))  #DASH bus stop layer

outfile=tempfile().  #create temp file on disk

py$map1$export_to_html(outfile) #write the html doc to temp file

htmltools::includeHTML(outfile) #this copys to memory and rendered in Rmd

file.remove(outfile)  #remove it right away
`

**Comment by gvwilson** [comment]
I'm generating GitHub Flavored Markdown from RMarkdown (because legacy websites...) and would like to have Plotly produce a .png on disk + a link in the generated .md the same way ggplot does when rendering with `md_document(...)` as the format.

**Comment by ichsan2895** [comment]
> Hello @dfrail24 , well I finally used the following approach:
> 
> * save the plotly graph made in Python as a JSON string
> 
> ```
>  ```{python}
> (...)
> fig = go.Figure(data=data, layout=layout)
> fig = json.dumps(fig, cls=plotly.utils.PlotlyJSONEncoder)
> ```
> 
> * read the JSON plot in R and plot it in R using R's plotly.
> 
> ```
> ```{r display_plotly_plot, message=FALSE, warning=FALSE}
> library(plotly)
> library(jsonlite)
> plotly::as_widget(
>   jsonlite::fromJSON(
>     reticulate::py$fig, simplifyVector=FALSE))
> ```
> 
> This way I was able to do the plot in Python, and use R Markdown to generate HTML slides. See the code at https://github.com/mwouts/jupytext_pyparis_2018/blob/master/slides/Jupytext.Rmd#L129-L180 and the resulting [HTML slides](https://mwouts.github.io/jupytext_pyparis_2018/slides/Jupytext.html#9).

Thanks, it worked for me

But I still hope, there are native support for python plotly in RStudio Rmarkdown

**Comment by cderv** [comment]
Just passing by and I think this may have been implemented now following discussion in https://github.com/rstudio/reticulate/issues/711

**Comment by mwouts** [comment]
I am not sure about `reticulate`, but I can confirm that Python Plotly plots now work very well within [Quarto](https://quarto.org/) (i.e `.qmd` extension rather than `.Rmd`), at least when exporting to `revealjs` HTML slides.

**Comment by kcarnold** [comment]
This works when Knit/Render-ing, but plots still show the spec rather than the plot in the IDE. See https://github.com/rstudio/rstudio/issues/8762 for a workaround.

(Maybe I'm dumb, but: why doesn't RStudio use the Jupyter wire protocol to talk to a Python interpreter? It handles all this stuff already.)

**Comment by t-kalinowski** [comment]
Plotly support in knitr chunks implemented with #711.
@kcarnold If there are friction points you're encountering, can you please open a new issue?

**Comment by kcarnold** [comment]
See kevinushey's comment at that issue: https://github.com/rstudio/reticulate/issues/711#issuecomment-1007006897

Technically there is no more bug in `reticulate`; it's the IDE now. But the vast majority of users (including me until a few days ago) will not realize that there's two different things, so they'll see the missing plot in the IDE and assume a reticulate bug. One simple workaround in the mean time would be to edit the description of this issue to include the text of kevinushey's comment.
